packages/gatsby/src/services/source-nodes.ts

```diff
@@ -0,0 +1,55 @@
+import { IBuildContext } from "../state-machines/develop"
+import sourceNodesAndGc from "../utils/source-nodes"
+import reporter from "gatsby-cli/lib/reporter"
+import { findChangedPages } from "../utils/check-for-changed-pages"
+
+export async function sourceNodes({
+  parentSpan,
+  webhookBody,
+  store,
+}: IBuildContext): Promise<{
+  changedPages: string[]
+  deletedPages: string[]
+} | void> {
+  if (!store) {
+    reporter.panic(`No redux store`)
+    return void 0
+  }
+  const activity = reporter.activityTimer(`source and transform nodes`, {
+    parentSpan,
+  })
+  console.log({ webhookBody })
+  activity.start()
+  const currentPages = new Map(store.getState().pages)
+  await sourceNodesAndGc({
+    parentSpan: activity.span,
+    deferNodeMutation: !!(webhookBody && Object.keys(webhookBody).length),
+    webhookBody,
+  })
+  reporter.info(`Checking for deleted pages`)
+
+  const tim = reporter.activityTimer(`Checking for changed pages`)
+  tim.start()
+
+  const { changedPages, deletedPages } = findChangedPages(
+    currentPages,
+    store.getState().pages
+  )
```

Comment:
I assume this is to handle case of creating / deleting pages in response to node creation/deletion ? Would be great if we could add some comments why we do anything about pages and "source-nodes"

Reply:
Yes, in this branch we allow pages to be created in sourceNodes, so we need to check here too